File: framec_tests/javascript/src/hierarchical/hierarchical.js
// emitted from framec_v0.10.0
// get include files at https://github.com/frame-lang/frame-ancillary-files

function FrameEvent(message, parameters) {

    var that = {};
    that._message = message;
    that._parameters = parameters;
    that._return = null;
    return that;
    
}

class Hierarchical {
    
    // creating private properties
    
    #state
    #compartment
    #nextCompartment
    
    constructor () {
        
        // Create and intialize start state compartment.
        
        this.#state = this.#sI_;
        this.#compartment = new HierarchicalCompartment(this.#state);
        this.#nextCompartment = null;
        
        // Initialize domain
        this.enters = [];
        this.exits = [];
        this.tape = [];
        
        // Send system start event
        const frameEvent = FrameEvent(">", null);
        this.#mux(frameEvent);
    }
    
    //===================== Interface Block ===================//
    
    A() {
        let e = FrameEvent("A",null);
        this.#mux(e);
    }
    
    B() {
        let e = FrameEvent("B",null);
        this.#mux(e);
    }
    
    C() {
        let e = FrameEvent("C",null);
        this.#mux(e);
    }
    
    //====================== Multiplexer ====================//
    
    #mux(e) {
        switch (this.#compartment.state) {
            case this.#sI_:
                this.#sI_(e);
                break;
            case this.#sS_:
                this.#sS_(e);
                break;
            case this.#sS0_:
                this.#sS0_(e);
                break;
            case this.#sS1_:
                this.#sS1_(e);
                break;
            case this.#sS2_:
                this.#sS2_(e);
                break;
            case this.#sS3_:
                this.#sS3_(e);
                break;
            case this.#sT_:
                this.#sT_(e);
                break;
        }
        
        if( this.#nextCompartment != null) {
            let nextCompartment = this.#nextCompartment
            this.#nextCompartment = null
            if (nextCompartment._forwardEvent != null && 
               nextCompartment._forwardEvent._message == ">") {
                this.#mux(FrameEvent( "<", this.#compartment.ExitArgs))
                this.#compartment = nextCompartment
                this.#mux(nextCompartment._forwardEvent)
            } else {
                this.#doTransition(nextCompartment)
                if (nextCompartment._forwardEvent != null) {
                    this.#mux(nextCompartment._forwardEvent)
                }
            }
            nextCompartment._forwardEvent = null
        }
    }
    
    //===================== Machine Block ===================//
    
    #sI_(e) {
        switch (e._message) {
            case ">":
                {
                let compartment =  new HierarchicalCompartment(this.#sS_);
                
                
                this.#transition(compartment);
                
                return;
                }
                
        }
    }
    
    #sS_(e) {
        switch (e._message) {
            case ">":
                {
                this.enter_do("S");
                
                return;
                }
                
            case "<":
                {
                this.exit_do("S");
                
                return;
                }
                
            case "A":
                {
                this.log_do("S.A");
                let compartment =  new HierarchicalCompartment(this.#sS0_);
                
                
                this.#transition(compartment);
                
                return;
                }
                
            case "B":
                {
                this.log_do("S.B");
                let compartment =  new HierarchicalCompartment(this.#sS1_);
                
                
                this.#transition(compartment);
                
                return;
                }
                
        }
    }
    
    #sS0_(e) {
        switch (e._message) {
            case ">":
                {
                this.enter_do("S0");
                
                break;
                }
                
            case "<":
                {
                this.exit_do("S0");
                
                break;
                }
                
              //  override parent handler
			case "A":
                {
                this.log_do("S0.A");
                let compartment =  new HierarchicalCompartment(this.#sT_);
                
                
                this.#transition(compartment);
                
                return;
                }
                
              //  do this, then parent handler
			case "B":
                {
                this.log_do("S0.B");
                
                break;
                }
                
              //  extend parent handler
			case "C":
                {
                this.log_do("S0.C");
                let compartment =  new HierarchicalCompartment(this.#sS2_);
                
                
                this.#transition(compartment);
                
                return;
                }
                
        }
        this.#sS_(e);
        
    }
    
    #sS1_(e) {
        switch (e._message) {
            case ">":
                {
                this.enter_do("S1");
                
                return;
                }
                
            case "<":
                {
                this.exit_do("S1");
                
                return;
                }
                
              //  defer to parent for A
			  //  do this, then parent, which transitions here
			case "B":
                {
                this.log_do("S1.B");
                
                break;
                }
                
              //  propagate message not handled by parent
			case "C":
                {
                this.log_do("S1.C");
                
                break;
                }
                
        }
        this.#sS_(e);
        
    }
    
    #sS2_(e) {
        switch (e._message) {
            case ">":
                {
                this.enter_do("S2");
                
                break;
                }
                
            case "<":
                {
                this.exit_do("S2");
                
                break;
                }
                
              //  will propagate to S0 and S
			case "B":
                {
                this.log_do("S2.B");
                
                break;
                }
                
            case "C":
                {
                this.log_do("S2.C");
                let compartment =  new HierarchicalCompartment(this.#sT_);
                
                
                this.#transition(compartment);
                
                return;
                break;
                }
                
        }
        this.#sS0_(e);
        
    }  //  continue after transition (should be ignored)
	
    
    #sS3_(e) {
        switch (e._message) {
            case ">":
                {
                this.enter_do("S3");
                
                break;
                }
                
            case "<":
                {
                this.exit_do("S3");
                
                break;
                }
                
              //  defer to grandparent for A
			  //  override and move to sibling
			case "B":
                {
                this.log_do("S3.B");
                let compartment =  new HierarchicalCompartment(this.#sS2_);
                
                
                this.#transition(compartment);
                
                return;
                }
                
        }
        this.#sS1_(e);
        
    }
    
    #sT_(e) {
        switch (e._message) {
            case ">":
                {
                this.enter_do("T");
                
                return;
                }
                
            case "<":
                {
                this.exit_do("T");
                
                return;
                }
                
            case "A":
                {
                this.log_do("T.A");
                let compartment =  new HierarchicalCompartment(this.#sS_);
                
                
                this.#transition(compartment);
                
                return;
                }
                
            case "B":
                {
                this.log_do("T.B");
                let compartment =  new HierarchicalCompartment(this.#sS2_);
                
                
                this.#transition(compartment);
                
                return;
                }
                
            case "C":
                {
                this.log_do("T.C");
                let compartment =  new HierarchicalCompartment(this.#sS3_);
                
                
                this.#transition(compartment);
                
                return;
                }
                
        }
    }
    
    //===================== Actions Block ===================//
    
    // Unimplemented Actions
    
    enter_do(msg) { throw new Error('Action not implemented.'); }
    exit_do(msg) { throw new Error('Action not implemented.'); }
    log_do(msg) { throw new Error('Action not implemented.'); }
    
    //=============== Machinery and Mechanisms ==============//
    
    #transition(compartment) {
        this.#nextCompartment = compartment;
    }
    
    #doTransition(nextCompartment) {
        this.#mux(FrameEvent("<", this.#compartment.ExitArgs));
        this.#compartment = nextCompartment;
        this.#mux(FrameEvent(">", this.#compartment.EnterArgs));
    }
    
    state_info() {
        return this.#compartment.state.name;
    }
    
    
};

//=============== Compartment ==============//

class HierarchicalCompartment {

    constructor(state) {
        this.state = state
    }
    
    StateArgs = {};
    StateVars = {};
    EnterArgs = {};
    ExitArgs = {};
    _forwardEvent = FrameEvent.call(this)
}


/********************

class HierarchicalController extends Hierarchical {

	constructor() {
	  super()
	}
	enter_do(msg) {}
	exit_do(msg) {}
	log_do(msg) {}
};

********************/

module.exports = Hierarchical

File: framec_tests/java/Hierarchical/Hierarchical.java
// emitted from framec_v0.10.0
// get include files at https://github.com/frame-lang/frame-ancillary-files

package framec_tests.java.Hierarchical;
import java.util.*;
import framec_tests.java.FrameLang.FrameEvent;

class Hierarchical {
    
    private HierarchicalCompartment _compartment_;
    private HierarchicalCompartment _nextCompartment_;
    
    
    
    Hierarchical() {
        
        // Create and intialize start state compartment.
        this._compartment_ = new HierarchicalCompartment(this._state_);
        this._nextCompartment_ = null;
        
        
        // Send system start event
        FrameEvent frame_event = new FrameEvent(">", null);
        this._mux_(frame_event);
    
    }
    
    // states enum
    private enum HierarchicalState {
        
        I(0),
        S(1),
        S0(2),
        S1(3),
        S2(4),
        S3(5),
        T(6);
        
        public final int value;
        
        private HierarchicalState(int value) {
            this.value=value;
        }
        
        public int getValue() {
            return value;
        }
    }
    
    //====================== Multiplexer ====================//
    
    private void _mux_(FrameEvent e) {
        if(this._compartment_.state == HierarchicalState.I.getValue()) {
            this._sI_(e);
        }else if(this._compartment_.state == HierarchicalState.S.getValue()) {
            this._sS_(e);
        }else if(this._compartment_.state == HierarchicalState.S0.getValue()) {
            this._sS0_(e);
        }else if(this._compartment_.state == HierarchicalState.S1.getValue()) {
            this._sS1_(e);
        }else if(this._compartment_.state == HierarchicalState.S2.getValue()) {
            this._sS2_(e);
        }else if(this._compartment_.state == HierarchicalState.S3.getValue()) {
            this._sS3_(e);
        }else if(this._compartment_.state == HierarchicalState.T.getValue()) {
            this._sT_(e);
        }
        
        
        if(this._nextCompartment_ != null) {
            HierarchicalCompartment nextCompartment = this._nextCompartment_;
            this._nextCompartment_ = null;
            if(nextCompartment._forwardEvent != null && 
             nextCompartment._forwardEvent._message == ">") {
                this._mux_(new FrameEvent( "<", this._compartment_.exitArgs));
                this._compartment_ = nextCompartment;
                this._mux_(nextCompartment._forwardEvent);
            } else {
                this._doTransition_(nextCompartment);
                if(nextCompartment._forwardEvent != null) {
                    this._mux_(nextCompartment._forwardEvent);
                }
            }
            nextCompartment._forwardEvent = null;
        }
    }
    
    //===================== Interface Block ===================//
    
    public void A() {
        FrameEvent e = new FrameEvent("A", null);
        this._mux_(e);
    }
    
    public void B() {
        FrameEvent e = new FrameEvent("B", null);
        this._mux_(e);
    }
    
    public void C() {
        FrameEvent e = new FrameEvent("C", null);
        this._mux_(e);
    }
    
    
    //===================== Machine Block ===================//
    
    private void _sI_(FrameEvent e) {
        if(e._message == ">") {
            HierarchicalCompartment compartment =  new HierarchicalCompartment(HierarchicalState.S.getValue());
            
            this._transition_(compartment);
            return;
        }
    }
    
    private void _sS_(FrameEvent e) {
        if(e._message == ">") {
            enter_do("S");
            return;
        }
        else if(e._message == "<") {
            exit_do("S");
            return;
        }
        else if(e._message == "A") {
            log_do("S.A");
            HierarchicalCompartment compartment =  new HierarchicalCompartment(HierarchicalState.S0.getValue());
            
            this._transition_(compartment);
            return;
        }
        else if(e._message == "B") {
            log_do("S.B");
            HierarchicalCompartment compartment =  new HierarchicalCompartment(HierarchicalState.S1.getValue());
            
            this._transition_(compartment);
            return;
        }
    }
    
    private void _sS0_(FrameEvent e) {
        if(e._message == ">") {
            enter_do("S0");
            
        }
        else if(e._message == "<") {
            exit_do("S0");
            
        }
          //  override parent handler
        else if(e._message == "A") {
            log_do("S0.A");
            HierarchicalCompartment compartment =  new HierarchicalCompartment(HierarchicalState.T.getValue());
            
            this._transition_(compartment);
            return;
        }
          //  do this, then parent handler
        else if(e._message == "B") {
            log_do("S0.B");
            
        }
          //  extend parent handler
        else if(e._message == "C") {
            log_do("S0.C");
            HierarchicalCompartment compartment =  new HierarchicalCompartment(HierarchicalState.S2.getValue());
            
            this._transition_(compartment);
            return;
        }
        _sS_(e);
        
    }
    
    private void _sS1_(FrameEvent e) {
        if(e._message == ">") {
            enter_do("S1");
            return;
        }
        else if(e._message == "<") {
            exit_do("S1");
            return;
        }
          //  defer to parent for A
          //  do this, then parent, which transitions here
        else if(e._message == "B") {
            log_do("S1.B");
            
        }
          //  propagate message not handled by parent
        else if(e._message == "C") {
            log_do("S1.C");
            
        }
        _sS_(e);
        
    }
    
    private void _sS2_(FrameEvent e) {
        if(e._message == ">") {
            enter_do("S2");
            
        }
        else if(e._message == "<") {
            exit_do("S2");
            
        }
          //  will propagate to S0 and S
        else if(e._message == "B") {
            log_do("S2.B");
            
        }
        else if(e._message == "C") {
            log_do("S2.C");
            HierarchicalCompartment compartment =  new HierarchicalCompartment(HierarchicalState.T.getValue());
            
            this._transition_(compartment);
            return;
        }
        _sS0_(e);
        
    }  //  continue after transition (should be ignored)
    
    
    private void _sS3_(FrameEvent e) {
        if(e._message == ">") {
            enter_do("S3");
            
        }
        else if(e._message == "<") {
            exit_do("S3");
            
        }
          //  defer to grandparent for A
          //  override and move to sibling
        else if(e._message == "B") {
            log_do("S3.B");
            HierarchicalCompartment compartment =  new HierarchicalCompartment(HierarchicalState.S2.getValue());
            
            this._transition_(compartment);
            return;
        }
        _sS1_(e);
        
    }
    
    private void _sT_(FrameEvent e) {
        if(e._message == ">") {
            enter_do("T");
            return;
        }
        else if(e._message == "<") {
            exit_do("T");
            return;
        }
        else if(e._message == "A") {
            log_do("T.A");
            HierarchicalCompartment compartment =  new HierarchicalCompartment(HierarchicalState.S.getValue());
            
            this._transition_(compartment);
            return;
        }
        else if(e._message == "B") {
            log_do("T.B");
            HierarchicalCompartment compartment =  new HierarchicalCompartment(HierarchicalState.S2.getValue());
            
            this._transition_(compartment);
            return;
        }
        else if(e._message == "C") {
            log_do("T.C");
            HierarchicalCompartment compartment =  new HierarchicalCompartment(HierarchicalState.S3.getValue());
            
            this._transition_(compartment);
            return;
        }
    }
    
    //===================== Actions Block ===================//
    
    // Unimplemented Actions
    
    protected void enter_do(String msg) { throw new UnsupportedOperationException(); }
    protected void exit_do(String msg) { throw new UnsupportedOperationException(); }
    protected void log_do(String msg) { throw new UnsupportedOperationException(); }
    
    //===================== Domain Block ===================//
    
    public ArrayList<String> enters  = new ArrayList<String>();
    public ArrayList<String> exits  = new ArrayList<String>();
    public ArrayList<String> tape  = new ArrayList<String>();
    
    //=============== Machinery and Mechanisms ==============//
    
    private int _state_;
    
    private void _transition_(HierarchicalCompartment compartment) {
        this._nextCompartment_ = compartment;
    }
    
    private void _doTransition_(HierarchicalCompartment nextCompartment) {
        this._mux_(new FrameEvent("<", this._compartment_.exitArgs));
        this._compartment_ = nextCompartment;
        this._mux_(new FrameEvent(">", this._compartment_.enterArgs));
    }
    
    public String state_info(){
        return String.valueOf(this._compartment_.state);
            }
            
    }
    
    //=============== Compartment ==============//
    
    class HierarchicalCompartment {
    
        int state;
        
        HierarchicalCompartment(int state) {
            this.state = state;
        }
        
        HashMap<String, Object> stateArgs = new HashMap<String, Object>();
        HashMap<String, Object> stateVars = new HashMap<String, Object>();
        HashMap<String, Object> enterArgs = new HashMap<String, Object>();
        HashMap<String, Object> exitArgs = new HashMap<String, Object>();
        FrameEvent _forwardEvent = new FrameEvent();
    }
    
    
    /********************

    class HierarchicalController extends Hierarchical {

    	HierarchicalController() {
    	  super();
    	}
    
    protected void enter_do(String msg) {}
    
    protected void exit_do(String msg) {}
    
    protected void log_do(String msg) {}
    }
    
********************/
    